### DISCUSSION TITLE
Restore teslamate backup when install with portainer

### DISCUSSION THREAD

**Discussion by w0rd3r** [teslamate-org/teslamate#2311] [Q&A] [answered]
Hello, 

If I decide to deploy Teslamate via portainer and its stack editor, I don't know how to then restore a Teslamate backup.

As a matter of fact, I cannot use this:
`# Stop the teslamate container to avoid write conflicts
docker-compose stop teslamate

# Drop existing data and reinitialize
docker-compose exec -T database psql -U teslamate << .
drop schema public cascade;
create schema public;
create extension cube;
create extension earthdistance;
CREATE OR REPLACE FUNCTION public.ll_to_earth(float8, float8)
    RETURNS public.earth
    LANGUAGE SQL
    IMMUTABLE STRICT
    PARALLEL SAFE
    AS 'SELECT public.cube(public.cube(public.cube(public.earth()*cos(radians(\$1))*cos(radians(\$2))),public.earth()*cos(radians(\$1))*sin(radians(\$2))),public.earth()*sin(radians(\$1)))::public.earth';
.

# Restore
docker-compose exec -T database psql -U teslamate -d teslamate < teslamate.bck

# Restart the teslamate container
docker-compose start teslamate`

since I did not use docker-compose.

Does anyone know how to drop data, reinitialize, and then restore if we did not use docker-compose to create Telamate?
Everything else works perfectly, and if I start again from scratch, put my file in a docker-compose.yml and then use docker-compose to create, drop, restore, it works.

I've tried using docker exec instead of docker-compose exec but can't seem to find the way.

Thanks!

**Comment by martinnj** [comment]
Hi!

Another edit: This is untested, btw.

Usually when starting containers in stacks (I only have experience with compose, but I suspect Portainer is the same) you can get some funky container names. When using `docker-compose exec` then stack name is often given as a project name or other argument, but for `docker exec` you will have to specify it directly in the container name.

```sh
$ docker exec teslamate_database_1 psql -U teslamate -d teslamate < teslamate.bck
```
Should probably do the trick, if it doesn't try using `docker ps` and check the name of the container (very right column, might get wrapped and end up on a second line).

EDIT:
If you mount the bck file inside the new database container, you can use Portainers "console" tab to run the restore command (`psql -U teslamate -d teslamate < teslamate.bck`) there, but it seems like a hassle to me. :)

**Comment by LukeSkywalker993** [comment, marked as answer]
1. backup on old server 
docker compose exec -T database pg_dump -U teslamate teslamate > ./teslamate.bck

1. Install teslamate on new server (with portainer)
2. check database name
cp to new server with the mc oder what ever

(check your database name!)
stop all services on old. stop all services on new, but not teslamate-database!
run:
docker exec -i teslamate-database-1 psql -U teslamate -d teslamate -c "DROP SCHEMA public CASCADE;"
docker exec -i teslamate-database-1 psql -U teslamate -d teslamate -c "CREATE SCHEMA public;"
docker exec -i teslamate-database-1 psql -U teslamate -d teslamate -c "CREATE EXTENSION cube WITH SCHEMA public;"
docker exec -i teslamate-database-1 psql -U teslamate -d teslamate -c "CREATE EXTENSION earthdistance WITH SCHEMA public;"

need time 10 GB about 15 min on fast server with nvme and xeon
docker exec -i teslamate-database-1 psql -U teslamate -d teslamate < teslamate.bck

restart all services

**Reply by Joe123-B** [reply]
thank you for the very nice instructions [LukeSkywalker993](https://github.com/LukeSkywalker993) - worked like a charme!

**Reply by dnuc** [reply]
Thanks @LukeSkywalker993 - best "tutorial" I found so far. Worked great!

**Reply by Polluxusss** [reply]
Due to the upgrade of PostgreSQL from v15 to v17 I had to do a backup and restore of my database in Portainer as well.
Thanks to [this post](https://github.com/teslamate-org/teslamate/issues/3518#issuecomment-1837199633) I was able to create a backup and @LukeSkywalker993 post helped me to restore it!
